File: moderate/alternative_reality.lua
m = {50, 25, 10, 5, 1}

function alter(n, p)
  if n == 0 then return 1 end
  while m[p] > n do p = p + 1 end
  if m[p] == 1 then return 1 end
  return alter(n - m[p], p) + alter(n, p + 1)
end

for line in io.lines(arg[1]) do
  print(alter(tonumber(line), 1))
end

File: moderate/alternative_reality.scala
object Main extends App {
  val source = scala.io.Source.fromFile(args(0))
  val lines = source.getLines.filter(_.length > 0)
  val m = List(50, 25, 10, 5, 1)

  def alter(n: Int, p: Int): Int = {
    if (n == 0) 1 else {
      var q = p
      while (m(q) > n) {
        q += 1
      }
      if (m(q) == 1) 1 else alter(n - m(q), q) + alter(n, q + 1)
    }
  }

  for (l <- lines)
    println(alter(l.toInt, 0))
}
